mitiq/zne/scaling/identity_insertion.py

```diff
@@ -0,0 +1,425 @@
+    folded circuit is approximately equal to scale_factor times the
+    total weight of the input circuit.
+    For equal weights, this function reproduces identity scaled version of
+    the local unitary folding method defined in equation (5) of
+    :cite:`Giurgica_Tiron_2020_arXiv`.
+    Args:
+        weight_mask: The weights of all the gates of the circuit to fold.
+            Highly noisy gates should have a corresponding high weight.
+            Gates with zero weight are assumed to be ideal and are not folded.
+        scale_factor: The effective noise scale factor.
+        scaling_method: A string equal to "at_random", or "from_left", or
+            "from_right". Determines the partial folding method described in
+            :cite:`Giurgica_Tiron_2020_arXiv`. If scale_factor is an odd
+            integer, all methods are equivalent and this option is irrelevant.
+        seed: A seed for the random number generator. This is used only when
+            scaling_method is "at_random".
+    Returns: The list of integers determining to how many times one should
+            insert I in the circuit to be scaled.
+    Example:
+        >>>_create_scale_mask(
+            weight_mask=[1.0, 0.5, 2.0, 0.0],
+            scale_factor=4,
+            scaling_method="from_left",
+        )
+        [2, 2, 1, 0]
+    """
+
+    if not 1.0 <= scale_factor:
+        raise ValueError(
+            f"Requires scale_factor >= 1.0 but scale_factor = {scale_factor}."
+        )
+
+    # Find the maximum odd integer smaller or equal to scale_factor
+    num_uniform_folds = int(scale_factor - 1.0)
+    odd_integer_scale_factor = np.abs(num_uniform_folds - 1)
+
+    # Uniformly scaling all gates to reach odd_integer_scale_factor
+    num_inserts_mask = []
+    for w in weight_mask:
+        if np.isclose(w, 0.0):
+            num_inserts_mask.append(0)
+        else:
+            num_inserts_mask.append(num_uniform_folds)
+
+    # If the scale_factor is an odd integer, we are done.
+    if np.isclose(odd_integer_scale_factor, scale_factor, atol=0.01):
+        return num_inserts_mask
+
+    # Express scaling order through a list of indices
+    scaling_order = list(range(len(weight_mask)))
+    if scaling_method == "from_left":
+        pass
+    elif scaling_method == "from_right":
+        scaling_order.reverse()
+    elif scaling_method == "at_random":
+        rnd_state = np.random.RandomState(seed)
+        rnd_state.shuffle(scaling_order)
+    else:
+        raise ValueError(
+            "The option 'scaling_method' is not valid."
+            "It must be 'at_random', or 'from_left', or 'from_right'."
+        )
+
+    # Fold gates until the input scale_factor is better approximated
+    input_circuit_weight = sum(weight_mask)
+    output_circuit_weight = odd_integer_scale_factor * input_circuit_weight
+    approx_error = np.abs(
+        output_circuit_weight - scale_factor * input_circuit_weight
+    )
```

Comment:
This is based on the previous suggested renaming.

```suggestion
    # Fold gates until the input scale_factor is better approximated
    input_circuit_weight = sum(weight_mask)
    output_circuit_weight = integer_scale_factor * input_circuit_weight
    approx_error = np.abs(
        output_circuit_weight - scale_factor * input_circuit_weight
    )
```